### ISSUE TITLE
Guardrails: Prevent writing when disk is almost exhausted 

### ISSUE THREAD

**Issue by tzach** [scylladb/scylladb#14067]
When storage is exhausted, it can be hard to fix, and even scaling or deleting becomes challenging.
It would be helpful to block writes *before* a node gets to this state.

This is similar to [Guardrails: Prevent writing to huge partitions #12995](https://github.com/scylladb/scylladb/issues/12995) but for all the tables (or at least the relevant keyspaces)

**Comment by nyh** [comment]
It makes sense not to reach the state of 100% disk full, but I wonder if it will really help anything in practice: If we block all writes on all replicas then basically we just killed the entire cluster; If we block writes on just one replica and not on the other replicas, we will lose write consistency in the same way that completely losing the node would have code. We do allow reads on this node - but they are all outdated and will also cause read repairs if using CL>1. Also, we should consider that not just user requests write to the disk - we also have hints (in case this node continues to act as a coordinator), read repairs (which we will need to not honor, but then read repairs will continue to happen again and again) and compactions (if we don't stop those too).

In https://github.com/scylladb/scylladb/issues/3524 I suggested a different way to handle the situation where we are nearing disk being full but not there yet - perhaps finding a way to recover space instead of blocking writes. Of course if the user continues to write new data eventually disk space will run out, so your idea will still be relevant in that case.

**Comment by tzach** [comment]
>It makes sense not to reach the state of 100% disk full, but I wonder if it will really help anything in practice: If we block all writes on all replicas then basically we just killed the entire cluster; 

Its not as bad as killing the cluster because you can still read, delete.
If the disk is 100% full, the cluster is in a much worse state, and it's much harder to recover (no space to compact, delete, etc)

#3524 is more civilized and can be used as the first line of defense, but if the user keeps adding new data, it will only delay the inevitable

**Comment by dorlaor** [comment]
I think it makes sense to fail writes when the node is almost full (95%),
the distance
from almost to completely full is small and it would be easier to recover
the node when
there is still some space left. It does create an out-of-sync problem with
the other replicas,
not a great situation but still better than 100% full.

Since this improvement is minor, I'm not sure if it's worth following

**Comment by vladzcloudius** [comment]
This proposal totally makes sense. 
I also agree with all @tzach's points from comment https://github.com/scylladb/scylladb/issues/14067#issuecomment-1566866823, in particular #3524 is orthogonal to this proposal. It completes it - not replaces it.

Some comments of my own: 

> It does create an out-of-sync problem with the other replicas, not a great situation but still better than 100% full.

Correct but to emphasize: if a node runs out of disk space and scylla goes down all writes to it are going to "fail" anyway so it's going to become "out-of-sync" regardless and a repair is going to be needed anyway.

But most importantly (what many people don't realize), the moment you loose a node you loose all its resources, and all that load is going to be redistributed across other nodes of the cluster, which in line can cause their overload and eventual total outage. 

And that's why we should do **everything** in our power to avoid that - even if the functionality that is left is partial. Like in this case, as Tzach correctly noted, the node can still serve reads as a replica and every request as a coordinator and this is a huge win over the current state of things.

And a final note about "out-of-sync" - you, guys, seem to forget that writes have a CL too. So, as long as we satisfy it - we are fine. And if CL is such that requires nodes with full disks to ack the write **you want** those CQL Writes to fail for the client - this is the fastest way for him/her to see the problem.

**Comment by nyh** [comment]
One small addition to what @vladzcloudius wrote about continuing to read from the (almost) "out of disk" node: We should modify the read-repair logic to not invoke read-repair when the differing node is one of those "out of disk" nodes. We *will* need to do the reconciliation of course, but not to send the writes of the reconciled data (those will fail on the "out of disk" node anyway, and will make everything slower instead of faster). 
But in some cases, I am afraid that even just the reconciliation algorithm - even without writing back the reconciled data - can mean that adding the almost-of-out-disk-node to reads can slow down everything instead of speeding it up. But I guess it depends on the exactly workload.

**Comment by vladzcloudius** [comment]
> One small addition to what @vladzcloudius wrote about continuing to read from the (almost) "out of disk" node: We should modify the read-repair logic to not invoke read-repair when the differing node is one of those "out of disk" nodes. We _will_ need to do the reconciliation of course, but not to send the writes of the reconciled data (those will fail on the "out of disk" node anyway, and will make everything slower instead of faster). But in some cases, I am afraid that even just the reconciliation algorithm - even without writing back the reconciled data - can mean that adding the almost-of-out-disk-node to reads can slow down everything instead of speeding it up. But I guess it depends on the exactly workload.

Correct. We definitely need to do all the instrumentation you have described before (https://github.com/scylladb/scylladb/issues/14067#issuecomment-1566853068).

The slow-down you described will only happen when the read row is inconsistent on the "read-only" node. How much such data is going to be there - it depends. How much such data is going to be read - also depends on the use case.

Its amount will obviously grow in time. However the assumption is that the operator is supposed to start fixing such an issue ASAP.

And I think it is totally acceptable if the performance degrades as long as such a "full" node is kept untreated. It's actually going to act similar to our MV flow control - the longer the backlog the higher is the base table write delay. ;)

**Comment by tzach** [comment]
Allowing DELETES is a double-edged sword; it *adds* to the storage first to free storage later during compaction.

**Comment by avikivity** [comment]
From #17729:

When free storage reaches min_free_storage_ratio (default 0.05 if tablets are enabled, otherwise 0.55), the replica rejects writes and stops writing hints. We can think about whether to allow it to compact or not. It still accepts writes from users that have the new WRITE WHEN OUT OF STORAGE permission. This can be used by the operator to delete data.

The node still responds to tablet migration requests (hopefully migrating out of the node), to group 0 writes (since they are needed for administration), and to reads.

**Comment by mykaul** [comment]
@bhalevy , @ptrsmrn - I suspect we'll critically need this at some point, I suggest revisiting this for 2025.2.

**Comment by bhalevy** [comment]
Cc @paszkow.This should be integrated with our write backpressure put-of-space prevention effort.

**Comment by ptrsmrn** [comment]
> Cc @paszkow.This should be integrated with our write backpressure put-of-space prevention effort.

@bhalevy your CC didn't work, I think.
On a side note, do I understand it correctly that guardrails effort is to be resumed? Or it's only about this specific one?

**Comment by bhalevy** [comment]
> ![Image](https://github.com/user-attachments/assets/2598593d-1d25-4bbc-8328-0cd1cec22b61)
> 
> > Cc @paszkow.This should be integrated with our write backpressure put-of-space prevention effort.
> 
> [@bhalevy](https://github.com/bhalevy) your CC didn't work, I think.

Yes, thanks.
Cc @paszkow 

> On a side note, do I understand it correctly that guardrails effort is to be resumed? Or it's only about this specific one?

It's about this specific one.
We have plans to apply write backpressure in response to nodes running out of space.  We wanted two thresholds: one to reject INSERT and UPDATE, but allow DELETE, and a higher one to reject all writes, including DELETE.
The mechanisms we had in mind would be implemented in lower layers in the stack, not in CQL front end, so we shouldn't hijack this issue for that.  With that in mind, I'm not sure what will be left to be done at the CQL level, if at all.
So @paszkow please have this issue in mind as part of the planning process.

**Comment by dorlaor** [comment]
These requirements are too complicated. Let's start by getting to agreement
in
google doc. Even delete needs compaction, it's not a realistic scenario to
separate the two
